### PR TITLE
feat: store and retrieve LTI launch data

### PR DESCRIPTION
**Pull request by pnal**
For feature https://github.com/oat-sa/extension-tao-outcome/pull/232:
- JSON serialization for AGS claims property in order to store LTILaunchData

### CODE REVIEW

File: models/classes/LtiLaunchData.php [hunk 1]
```diff
@@ -146,6 +150,7 @@ public static function fromLti1p3MessagePayload(
         LtiMessagePayloadInterface $ltiMessagePayload,
         PlatformInterface $platform = null
     ): self {
+        // phpcs:disable Generic.Files.LineLength
```
Comment (bartlomiejmarszal) [inline models/classes/LtiLaunchData.php:153]: Why you do not refactor this?

File: models/classes/LtiLaunchData.php [hunk 2]
```diff
@@ -550,7 +560,9 @@ public function getReturnUrl()
     public function jsonSerialize()
     {
         return [
-            'variables' => $this->variables,
+            'variables' => array_map(function ($var) {
+                return $var instanceof AgsClaim ? $var->normalize() : $var;
+            }, $this->variables),
```
Comment (bartlomiejmarszal) [inline models/classes/LtiLaunchData.php:565]: This is not good place for this operation. Should we extract it to some method instead?

File: test/unit/models/classes/LtiLaunchDataTest.php [hunk 3]
```diff
@@ -43,17 +44,51 @@ public function testFromRequest()
...
-            ->getMock();
-        $request->method('getUrl')->willReturn($url);
-        $request->method('getParams')->willReturn($params);
+        $request = $this->createConfiguredMock(
+            Request::class,
+            [
+                'getUrl' => $url,
+                'getParams' => $params
+            ]
+        );
```
Comment (bartlomiejmarszal) [inline test/unit/models/classes/LtiLaunchDataTest.php:53]: Nice one!

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
In jsonSerialise for better readability I recommend extracting denormalisation  process to private method.

File: models/classes/LtiLaunchData.php [hunk 4]
```diff
@@ -550,7 +560,9 @@ public function getReturnUrl()
     public function jsonSerialize()
     {
         return [
-            'variables' => $this->variables,
+            'variables' => array_map(function ($var) {
```
Comment (andreluizmachado) [inline models/classes/LtiLaunchData.php:563]: you can use arrow functions, no?

File: test/unit/models/classes/LtiLaunchDataTest.php [hunk 5]
```diff
@@ -43,17 +44,51 @@ public function testFromRequest()
...
 
         $subject = LtiLaunchData::fromRequest($request);
 
         $this->assertEquals($params, $subject->getVariables());
         $this->assertEquals($extraParams, $subject->getCustomParameters());
     }
+
+    public function testFromJsonArray()
```
Comment (andreluizmachado) [inline test/unit/models/classes/LtiLaunchDataTest.php:61]: return types

File: models/classes/LtiLaunchData.php (hunk 1 above)
Comment (pnal) [reply models/classes/LtiLaunchData.php:149]: too much to just simply break into multiple lines, do not see much sense in it,
at once no general idea how to refactor it properly, do you have one?

File: models/classes/LtiLaunchData.php (hunk 4 above)
Comment (pnal) [reply models/classes/LtiLaunchData.php]: used arrow fn

File: models/classes/LtiLaunchData.php (hunk 2 above)
Comment (pnal) [reply models/classes/LtiLaunchData.php]: Changed callback to arrow fn for better readability as @andreluizmachado proposed.
Anyway for me it looks much more transparent then moved out somewhere

File: models/classes/LtiLaunchData.php (hunk 1 above)
Comment (Karol-Stelmaczonek) [reply models/classes/LtiLaunchData.php:149]: Just split it in more lines like on image (couldn write it in markdonw :| ) 
image

**Review by Karol-Stelmaczonek — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: models/classes/LtiLaunchData.php [hunk 6]
```diff
@@ -108,12 +109,11 @@ public function __construct(array $ltiVariables, array $customParameters)
...
-    /**
-     * @param array $json
-     * @return LtiLaunchData
-     */
-    public static function fromJsonArray($json)
+    public static function fromJsonArray(array $json): LtiLaunchData
     {
+        if (isset($json['variables'][self::AGS_CLAIMS])) {
+            $json['variables'][self::AGS_CLAIMS] = AgsClaim::denormalize($json['variables'][self::AGS_CLAIMS]);
+        }
```
Comment (bartlomiejmarszal) [inline models/classes/LtiLaunchData.php:116]: extract this to private method

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by andreluizmachado — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: models/classes/LtiLaunchData.php [hunk 7]
```diff
@@ -143,63 +143,90 @@ public static function fromRequest(common_http_Request $request)
...
-
-        $variables[self::LTI_FOR_USER_ID] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getIdentifier() : null;
-        $variables[self::LTI_FOR_USER_EMAIL] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getEmail() : null;
-        $variables[self::LTI_FOR_USER_FAMILY_NAME] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getFamilyName() : null;
-        $variables[self::LTI_FOR_USER_GIVEN_NAME] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getGivenName() : null;
-        $variables[self::LTI_FOR_USER_NAME] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getName() : null;
-        $variables[self::LTI_FOR_USER_PERSON_SOURCED_ID] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getPersonSourcedId() : null;
-        $variables[self::LTI_FOR_USER_ROLES] = $ltiMessagePayload->getForUser() ? $ltiMessagePayload->getForUser()->getRoles() : null;
+        $variables[self::RESOURCE_LINK_ID] =
+            $payload->getResourceLink() ? $payload->getResourceLink()->getIdentifier() : null;
```
Comment (bartlomiejmarszal) [inline models/classes/LtiLaunchData.php:151]:
```suggestion
        $variables[self::RESOURCE_LINK_ID] = $payload->getResourceLink() 
          ? $payload->getResourceLink()->getIdentifier() 
          : null;
```

File: models/classes/LtiLaunchData.php [hunk 8]
```diff
@@ -143,63 +143,90 @@ public static function fromRequest(common_http_Request $request)
     }
 
     public static function fromLti1p3MessagePayload(
-        LtiMessagePayloadInterface $ltiMessagePayload,
+        LtiMessagePayloadInterface $payload,
```
Comment (bartlomiejmarszal) [inline models/classes/LtiLaunchData.php:146]: Why did you change that variable name? It can cause more harm then good.

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
Do we really need to rename ltiMessagePayload to payload?

File: models/classes/LtiLaunchData.php (hunk 8 above)
Comment (pnal) [reply models/classes/LtiLaunchData.php:146]: it is shorter and still meaningful cause has type hint
